ambry-cloud/src/main/java/com/github/ambry/cloud/azure/ADAuthBasedStorageClientFactory.java

```diff
@@ -0,0 +1,90 @@
+/**
+ * Copyright 2020 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.cloud.azure;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.core.http.HttpClient;
+import com.azure.core.util.Configuration;
+import com.azure.storage.blob.BlobServiceClient;
+import com.azure.storage.blob.BlobServiceClientBuilder;
+import com.azure.storage.common.policy.RequestRetryOptions;
+import com.microsoft.aad.msal4j.ClientCredentialFactory;
+import com.microsoft.aad.msal4j.ClientCredentialParameters;
+import com.microsoft.aad.msal4j.ConfidentialClientApplication;
+import com.microsoft.aad.msal4j.IAuthenticationResult;
+import java.net.MalformedURLException;
+import java.time.OffsetDateTime;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import reactor.core.publisher.Mono;
+
+
+/**
+ * {@link StorageClientFactory} implementation for AD based authentication.
+ */
+public class ADAuthBasedStorageClientFactory extends StorageClientFactory {
+  private static String AZURE_STORAGE_ACCESS_SCOPE = "https://wus2ambryblobstore1.blob.core.windows.net/.default";
+
+  @Override
+  protected BlobServiceClient buildBlobServiceClient(HttpClient httpClient, Configuration configuration,
+      RequestRetryOptions retryOptions, AzureCloudConfig azureCloudConfig)
+      throws MalformedURLException, InterruptedException, ExecutionException {
+    IAuthenticationResult iAuthenticationResult = getAccessTokenByClientCredentialGrant(azureCloudConfig);
+    return new BlobServiceClientBuilder().credential(new TokenCredential() {
```

Comment:
Minor: it will be easier to read if you split the credential construction into its own line.

Reply:
done.